src/CommunityCommons/javasource/communitycommons/actions/executeMicroflowAsSystemUser.java

```diff
@@ -0,0 +1,76 @@
+// This file was generated by Mendix Modeler.
+//
+// WARNING: Only the following code will be retained when actions are regenerated:
+// - the import list
+// - the code between BEGIN USER CODE and END USER CODE
+// - the code between BEGIN EXTRA CODE and END EXTRA CODE
+// Other code you write will be lost the next time you deploy the project.
+// Special characters, e.g., é, ö, à, etc. are supported in comments.
+
+package communitycommons.actions;
+
+import com.mendix.core.Core;
+import com.mendix.systemwideinterfaces.core.IContext;
+import com.mendix.webui.CustomJavaAction;
+import com.mendix.systemwideinterfaces.core.IMendixObject;
+
+/**
+ * This action will invoke a microflow, optionally with an argument and/or return value.
+ * It will execute the microflow in the system context.
+ * This is useful when performing a lot of changes and avoiding the overhead of the user context which will send updates/changes back to the client.
+ * A must have for > 100k updates.
+ */
+public class executeMicroflowAsSystemUser extends CustomJavaAction<IMendixObject>
+{
+	private java.lang.String microflow;
+	private java.lang.String microflowArgumentEntity;
+	private IMendixObject microflowArgument;
+	private java.lang.String microflowReturnType;
+
+	public executeMicroflowAsSystemUser(IContext context, java.lang.String microflow, java.lang.String microflowArgumentEntity, IMendixObject microflowArgument, java.lang.String microflowReturnType)
+	{
+		super(context);
+		this.microflow = microflow;
+		this.microflowArgumentEntity = microflowArgumentEntity;
+		this.microflowArgument = microflowArgument;
+		this.microflowReturnType = microflowReturnType;
+	}
+
+	@java.lang.Override
+	public IMendixObject executeAction() throws Exception
+	{
+		// BEGIN USER CODE
+		IMendixObject result = null;
+		IContext sysContext = Core.createSystemContext();
+		sysContext.startTransaction();
+		try {
+			Object r;
+			if (this.microflowArgument != null) {
+				r = Core.execute(sysContext, microflow, microflowArgument);
+			} else {
+				r = Core.execute(sysContext, microflow); 
+			}
+			if (r != null && r instanceof IMendixObject) {
+				result = (IMendixObject) r;
+			}
+		} finally {
+			while (sysContext.isInTransaction()) {
+				sysContext.endTransaction();
```

Comment:
The while-loop suggests that the context could still be 'in transaction' and that a single `endTransaction()` call won't have the desired effect. But if you keep calling `endTransaction()` in a while-loop and the transaction for some reason is not ending, the thread will be stuck in an endless loop.
So, 1) is it necessary to use a while-loop here instead of `if`? and 2) if yes, let's add a delay of 0.5sec or so in between calls to endTransaction().

Reply:
endTransaction closes the level of the transaction in progress. This can be the main transaction, a savepoint or a savepoint in a savepoint, etc.; hence the while construction which makes sure there's no open transaction left.